crates/driver/src/domain/competition/order/mod.rs

```diff
@@ -133,20 +133,57 @@ impl Order {
         matches!(self.kind, Kind::Liquidity)
     }
 
-    /// The buy asset to pass to the solver. This is a special case due to
-    /// orders which buy ETH. The settlement contract only works with ERC20
-    /// tokens, but unfortunately ETH is not an ERC20 token. We still want to
-    /// provide a seamless user experience for ETH trades, so the driver
-    /// will encode the settlement to automatically unwrap the WETH into ETH
-    /// after the trade is done.
+    /// The sell asset to pass to the solver.
     ///
-    /// For this reason, we want the solvers to solve the orders which buy ETH
-    /// as if they were buying WETH, and then add our unwrap interaction to that
-    /// solution.
+    /// In order to simplify solver logic, we scale the remaining sell amount
+    /// for orders that have been partially filled to simplify computations on
+    /// the solver side. This is so the solvers only see "whats left".
+    pub fn solver_sell(&self) -> eth::Asset {
+        self.scale_remaining(eth::Asset {
+            amount: self.sell.amount,
+            token: self.sell.token,
+        })
+    }
+
+    /// The buy asset to pass to the solver. This is a special case in two ways:
+    ///
+    /// 1. orders which buy ETH: The settlement contract only works with ERC20
+    ///    tokens, but unfortunately ETH is not an ERC20 token. We still want to
+    ///    provide a seamless user experience for ETH trades, so the driver will
+    ///    encode the settlement to automatically unwrap the WETH into ETH after
+    ///    the trade is done. For this reason, we want the solvers to solve the
+    ///    orders which buy ETH as if they were buying WETH, and then add our
+    ///    unwrap interaction to that solution.
+    /// 2. partial fill orders: The driver will scale the remaining amounts for
+    ///    orders that have been partially filled to simplify computations on
+    ///    the solver side. This is so the solvers only see "whats left".
     pub fn solver_buy(&self, weth: eth::WethAddress) -> eth::Asset {
-        eth::Asset {
+        self.scale_remaining(eth::Asset {
             amount: self.buy.amount,
             token: self.buy.token.wrap(weth),
+        })
+    }
+
+    /// Scales a full asset amount to how much is remaining in the order. That
+    /// is, if an order is 75% filled, then this method will return an asset
+    /// with the same token as `full` but with 25% of the amount.
+    fn scale_remaining(&self, full: eth::Asset) -> eth::Asset {
+        let Partial::Yes { executed } = self.partial else {
+            return full;
+        };
+
+        let target = self.target();
+        let remaining = target.0.saturating_sub(executed.0);
+        let amount = remaining
+            .full_mul(full.amount.0)
+            .checked_div(target.0.into())
+            .unwrap_or_default()
```

Comment:
Do we want to unwrap or default here rather than rounding `target.0` to 1 for the division? Would it make sense to have a unit test how this scaling works in edge cases?

Reply:
I chose to `unwrap_or_default` to make this function non-failing/non-panicking.

Reply:
My main worry is about edge cases like having a partially fillable limit order that is selling 1000 ETH for 1 wei of USDC. If this were to get executed to 50% in the first batch, would the remaining amount scaled cause us to have a 0 buy amount and therefore censor the order?

Reply:
Changed the `solver_buy` to do a ceiling division.